intermediate_source/rpc_tutorial.rst

```diff
@@ -0,0 +1,529 @@
+borrowed from the word language model in PyTorch
+`example <https://github.com/pytorch/examples/tree/master/word_language_model>`__
+repository, which contains three main components, an embedding table, an
+``LSTM`` layer, and a decoder, as shown below.
+
+
+.. code:: python
+
+    class EmbeddingTable(nn.Module):
+        def __init__(self, ntoken, ninp, dropout):
+            super(EmbeddingTable, self).__init__()
+            self.drop = nn.Dropout(dropout)
+            self.encoder = nn.Embedding(ntoken, ninp)
+            self.encoder.weight.data.uniform_(-0.1, 0.1)
+
+        def forward(self, input):
+            return self.drop(self.encoder(input))
+
+
+    class RNN(nn.Module):
+        def __init__(self, ninp, nhid, nlayers, dropout):
+            super(RNN, self).__init__()
+            self.lstm = nn.LSTM(ninp, nhid, nlayers, dropout=dropout)
+
+        def forward(self, emb, hidden):
+            return self.lstm(emb, hidden)
+
+
+    class Decoder(nn.Module):
+        def __init__(self, ntoken, nhid, dropout):
+            super(Decoder, self).__init__()
+            self.drop = nn.Dropout(dropout)
+            self.decoder = nn.Linear(nhid, ntoken)
+            self.decoder.bias.data.zero_()
+            self.decoder.weight.data.uniform_(-0.1, 0.1)
+
+        def forward(self, output):
+            return self.decoder(self.drop(output))
+
+
+With the above three sub-modules, we can now piece them together using RPC to
+create an RNN model. In the code below ``ps`` represents a parameter server,
+which hosts paremeters of the embedding table and the decoder. The constructor
+uses the `remote https://pytorch.org/docs/master/rpc.html#torch.distributed.rpc.remote`__
+API to create an `EmbeddingTable` and a `Decoder` object on the parameter
+server, and locally creates the ``LSTM`` sub-module. During the forward pass,
+the trainer uses the ``EmbeddingTable`` ``RRef`` to find the remote sub-module
+and passes the input data to the ``EmbeddingTable`` using RPC and fetches the
+lookup results. Then, it runs the embedding through the local ``LSTM`` layer,
+and finally uses another RPC to send the output to the ``Decoder`` sub-module.
+In general, to implement distributed model parallel training, developers can
+divide the model into sub-modules, invoke RPC to create sub-module instances
+remotely, and use on ``RRef`` to find them when necessary. As you can see in the
+code below, it looks very similar to single-machine model parallel training. The
+main difference is replacing ``Tensor.to(device)`` with RPC functions.
+
+
+.. code:: python
+
+    class RNNModel(nn.Module):
+        def __init__(self, ps, ntoken, ninp, nhid, nlayers, dropout=0.5):
+            super(RNNModel, self).__init__()
+
+            # setup embedding table remotely
+            self.emb_table_rref = rpc.remote(ps, EmbeddingTable, args=(ntoken, ninp, dropout))
+            # setup LSTM locally
+            self.rnn = nn.LSTM(ninp, nhid, nlayers, dropout=dropout)
```

Comment:
There is a `class RNN` defined above, should we remove that or use it here?

Reply:
Oh yes, good catch, let me remove RNN class to make it shorter